phoenix-core/src/main/java/org/apache/phoenix/query/ConnectionQueryServicesImpl.java

```diff
@@ -4957,12 +4964,22 @@ private void returnAllSequences(ConcurrentMap<SequenceKey,Sequence> sequenceMap)
     public void addConnection(PhoenixConnection connection) throws SQLException {
         if (returnSequenceValues || shouldThrottleNumConnections) {
             synchronized (connectionCountLock) {
-                if (shouldThrottleNumConnections && connectionCount + 1 > maxConnectionsAllowed){
-                    GLOBAL_PHOENIX_CONNECTIONS_THROTTLED_COUNTER.increment();
-                    throw new SQLExceptionInfo.Builder(SQLExceptionCode.NEW_CONNECTION_THROTTLED).
-                        build().buildException();
+
+                if(shouldThrottleNumConnections) {
+                    int futureConnections = 1 + ( connection.isInternalConnection() ? internalConnectionCount : connectionCount);
+                    int allowedConnections = connection.isInternalConnection() ? maxInternalConnectionsAllowed : maxConnectionsAllowed;
+                    if(allowedConnections != 0 && futureConnections > allowedConnections) {
+                        GLOBAL_PHOENIX_CONNECTIONS_THROTTLED_COUNTER.increment();
+                        throw new SQLExceptionInfo.Builder(SQLExceptionCode.NEW_CONNECTION_THROTTLED).
```

Comment:
Should we throw a different error for cases where internal conns exceed the max config vs. when external do?

Reply:
We could but I'm not sure there is much benefit as we already get different line number and the end client can't really act differently.

Reply:
I can see a different error message being useful, because it keeps the client developer from wasting a lot of time looking for a non-existent leak on their side.

Reply:
Added another sql exception code and surface different exception based on connection type.

Reply:
@dbwong I don't see this change. Can you please check again?